File: packages/sdk-ts/src/adapters/openai-agents.ts
/**
 * OpenAI Agents SDK (JavaScript) — `@openai/agents`. Tools built with `tool({ name, execute })` expose
 * `invoke(runContext, input)`; the guarded copy verifies before the original runs. Type-only: the framework
 * is never imported, so this file loads without it.
 *
 *   import { guardTools } from "mnki-sdk/openai-agents";
 *   const agent = new Agent({ name: "invoice-agent", tools: guardTools(guard, [refundTool, lookupTool]) });
 */
import { guarded, parseJsonArgs, wanted, type AdapterOptions, type GuardLike } from "./shared";

/** The subset of `FunctionTool` the adapter relies on. */
export interface OpenAIAgentsTool { name: string; invoke: (runContext: unknown, input: string, ...rest: unknown[]) => Promise<unknown> }

export function guardTool<T extends OpenAIAgentsTool>(guard: GuardLike, tool: T, o: AdapterOptions = {}): T {
  if (!wanted(tool.name, o)) return tool;
  const name = o.toolName ? o.toolName(tool.name) : tool.name;
  const original = tool.invoke.bind(tool);
  const invoke: OpenAIAgentsTool["invoke"] = async (runContext, input, ...rest) => {
    const r = await guarded(guard, name, parseJsonArgs(input), () => original(runContext, input, ...rest), o);
    // A refusal is returned to the model as a JSON string, the shape every function tool result takes.
    return typeof r === "object" && r !== null && "error" in (r as object) && "reasons" in (r as object) ? JSON.stringify(r) : r;
  };
  return Object.assign(Object.create(Object.getPrototypeOf(tool) as object), tool, { invoke }) as T;
}
export const guardTools = <T extends OpenAIAgentsTool>(guard: GuardLike, tools: T[], o: AdapterOptions = {}): T[] => tools.map((t) => guardTool(guard, t, o));

File: python/mnki/adapters/openai_agents.py
"""
OpenAI Agents SDK (Python) — `openai-agents`. `FunctionTool.on_invoke_tool(ctx, input_json)` is verified before
the original runs; every other attribute (name, params_json_schema, strict…) is untouched, so `Agent(tools=…)`
accepts the guarded copies.

    from agents import Agent, function_tool
    from mnki.adapters.openai_agents import guard_tools
    agent = Agent(name="invoice-agent", tools=guard_tools(guard, [refund, lookup]))
"""
from __future__ import annotations
import copy, json
from typing import Any, Optional

from .shared import check_or_refuse, parse_json_args, wanted


def guard_tool(guard: Any, tool: Any, on_refused: str = "result", only: Optional[list] = None, skip: Optional[list] = None, tool_name=None) -> Any:
    name = getattr(tool, "name", None) or getattr(tool, "__name__", "tool")
    if not wanted(name, only, skip): return tool
    verified = tool_name(name) if tool_name else name
    original = tool.on_invoke_tool

    async def on_invoke_tool(ctx: Any, input_json: str) -> Any:
        r = check_or_refuse(guard, verified, parse_json_args(input_json), on_refused)
        if r is not None: return json.dumps(r)
        return await original(ctx, input_json)

    g = copy.copy(tool)
    try: object.__setattr__(g, "on_invoke_tool", on_invoke_tool)
    except (AttributeError, TypeError): g = _replace(tool, on_invoke_tool)
    return g


def _replace(tool: Any, fn: Any) -> Any:
    import dataclasses
    if dataclasses.is_dataclass(tool): return dataclasses.replace(tool, on_invoke_tool=fn)
    raise TypeError("tool does not expose a writable on_invoke_tool")


def guard_tools(guard: Any, tools: list, **kw: Any) -> list: return [guard_tool(guard, t, **kw) for t in tools]
